### DISCUSSION TITLE
How can I make AFL++ allow to use llvm-symbolizer?

### DISCUSSION THREAD

**Discussion by junwha** [AFLplusplus/AFLplusplus#1618] [Q&A] [answered]
Hi guys,
Currently I'm trying to use modified Address sanitizer with utilizing the llvm-symbolizer related feature.
But I figured out that inside AFL++, the llvm-symbolizer is not working correctly.
For example, if I run the afl-fuzz with crash input, (also, with AFL_DEBUG=1)
![image](https://user-images.githubusercontent.com/17183234/212961714-9d5d247d-4364-4c32-850f-ba4eb549926d.png)
The llvm-symbolizer does not work correctly.

As I know, AFL++ automatically set the ASAN_OPTIONS which contains `symbolize=0`,
so tried again with `ASAN_OPTIONS=symbolize=0:abort_on_error=1`, but I got the same result.

I really want to use llvm-symbolizer with AFL++ because I need to automatically check the call stack in my tool.

Is there anyone can help me? 

Thank you.

**Comment by vanhauser-thc** [comment, marked as answer]
why would you need the symbolizer when fuzzing? did you make your own modifications to afl++?
outside of afl-fuzz you can set any asan options you want, also symbolize=1 which is a very good idea to follow up found crashes.

**Reply by junwha** [reply]
Yes, I already got that point. As you mentioned, we don't need to enable llvm-symbolizer inside AFL++ in normal case. 
But In my case, I use llvm-symbolizer in my own llvm pass, and compiler-rt will invoke llvm-symbolizer while it is running inside AFL++.
However, currently it fails to invoke llvm-symbolizer inside AFL++, and I like to know what is the reason for that.
Could I know how can I make my own modifications to AFL++ to be able to invoke llvm-symbolizer inside AFL++?

+I made a typo. I failed to enable llvm-symbolizer when symbolize=1 inside afl-fuzz.

Thank you.

**Reply by vanhauser-thc** [reply]
well if you write your own plugin and you need symbols from asan, then sure just remove the check in `src/afl-fuzz-init.c`

**Reply by junwha** [reply]
you mean AFL++ doesn't forcefully disable llvm-symbolizer? I had tried to ignore the warning from AFL++ and set ASAN_OPTIONS as symbolize=1:abort_on_error=1 before, but it show the same result as above. it runs, but symbolizer couldn't resolve the symbol of the functions. I thought that this is because AFL++ internally disable llvm-symbolizer, isn't there any implementation that suppresses llvm-symbolizer from working correctly?

**Reply by vanhauser-thc** [reply]
No it doesnt. just look at the code.
it just terminates with an error if it sees symbolize=1 and AFL_DEBUG is unset.

**Reply by junwha** [reply]
I see. I thought I missed some part of AFL++, but according to your answer, it seems it's my local problem. I'll try to find a problem. 
Thank you much!

**Reply by junwha** [reply]
I'm sorry to bother you. But as I think, there is something that suppress AFL++ to enable llvm-symbolizer other than the simple termination code. 
https://github.com/AFLplusplus/AFLplusplus/blob/stable/src/afl-fuzz-init.c#L2499
I think this line is what you've mentioned, and I couldn't see that message when I set the ASAN_OPTIONS with symbolize=1 because I enabled AFL_DEBUG=1.

Here is my simple buggy code experimented on the docker image `aflplusplus/aflplusplus:stable`.
```
#include<stdlib.h>
#include<stdio.h>
int main() {
        int *a = (int*) malloc(sizeof(int)*10);
        int idx;
        scanf("%d", &idx);
        printf("idx: %d\n", idx);
        a[idx] = 39877896;
        free(a);
        if (idx >= 10)
        printf("%d\n", a[idx]);
        return 0;
}
```
if input index is greater than 10, address sanitizer will detect heap buffer overflow from this example.
I compiled this with `afl-clang-fast bof.c -fsanitize=address -g`
(the other weird point is that, it was not enabled Address sanitizer when I set AFL_USE_ASAN=1)

I made 3, 9, 10 as input, and set `ASAN_OPTIONS=abort_on_error=1:symbolize=1` and `AFL_DEBUG=1`. 
Then, I run the fuzzer.
`afl-fuzz -i input -o output -- ./a.out`

According to what you told me, there must be normally working llvm-symbolizer, but it shows following result.
<img width="990" alt="image" src="https://user-images.githubusercontent.com/17183234/213905069-d9428bd9-4ff6-4768-b672-143837d6b3fa.png">

I didn't touched both LLVM and AFL++ inside the deployed docker image, but llvm-symbolizer didn't work. (also, I enrolled to add /usr/lib/llvm-14/bin to PATH and tried again, but same result was there)

Can I get some advice on what you think the problem is?

**Reply by vanhauser-thc** [reply]
I think its a behaviour in asan. 
if you check /proc/fuzzed_pid/environ while fuzzing you will see that ASAN_OPTIONS is untouched. so if asan doesnt show the lines in the source that is a decision within asan for a reason I don't know.

**Reply by junwha** [reply]
It was not the behaviour in asan, but it was sanitizer flag overwritten problem.

in [src/afl-forkserver.c](https://github.com/AFLplusplus/AFLplusplus/blob/3b6fcd911a860a8c823c912c4b08b423734e4cfe/src/afl-forkserver.c#L693), afl++ set abort_on_error and symbolize for all the four sanitizer flags, ASAN_OPTIONS, LSAN_OPTIONS, MSAN_OPTIONS, and UBSAN OPTIONS. 

And it is problematic because the four sanitizers internally share the [sanitizer common options](https://github.com/google/sanitizers/wiki/SanitizerCommonFlags). For this reason, my ASAN_OPTION symbolize=0 has been overwritten by three sanitizer's other options.
 
In my case, symbolize=1 was discouraged by AFL++,  thus it can be considered as tirvial problem. 
But I think this implementation is something that needs to be fixed because what the user has set up in one sanitizer has a potential risk of overwriting by AFL++ initialization.

Thank you again for your help.
